secp256k1_test.go

```diff
@@ -0,0 +1,534 @@
+	}
+
+	if *pubkey != originalPubKey {
+		t.Errorf("resulting privkey: %v doesn't match original: %v", privkey, originalPrivatekey)
+	}
+}
+
+func TestParseSchnorrPubKeyFail(t *testing.T) {
+	zeros := [65]byte{}
+	_, err := ParseSchnorrPubKey(zeros[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse 65 zeros as a pubkey %v", zeros)
+	}
+	_, err = ParseSchnorrPubKey(zeros[:33])
+	if err == nil {
+		t.Errorf("Shouldn't parse 33 zeros as a pubkey %v", zeros[:33])
+	}
+	zeros[0] = 0x04
+	_, err = ParseSchnorrPubKey(zeros[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse zeroed x and y as pubkey %v", zeros)
+	}
+	pubkey, err := fastGeneratePrivateKey(t).GenerateSchnorrPublicKey()
+	if err != nil {
+		t.Fatalf("Failed generating a random pubkey: %v", err)
+	}
+	compressed, err := pubkey.SerializeCompressed()
+	if err != nil {
+		t.Fatalf("Failed serializing a pubkey: %v %v", err, pubkey)
+	}
+	oddnessByte := compressed[0]
+	compressed[0] = 0x07
+	_, err = ParseSchnorrPubKey(compressed[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse a compressed key starting with 0x07 %v", compressed)
+	}
+	uncompressed, err := pubkey.SerializeUncompressed()
+	if err != nil {
+		t.Fatalf("Failed serializing a pubkey: %v %v", err, pubkey)
+	}
+	uncompressed[0] = oddnessByte
+	_, err = ParseSchnorrPubKey(uncompressed[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse an uncompressed key starting with 0x2/0x03 %v", uncompressed)
+	}
+	uncompressed[0] = oddnessByte + 0x04
+	_, err = ParseSchnorrPubKey(uncompressed[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse a weird hybrid pubkey %v", uncompressed)
+	}
+	uncompressed[0] = 0x04
+	uncompressed[64] += 0x01
+	_, err = ParseSchnorrPubKey(uncompressed[:])
+	if err == nil {
+		t.Errorf("Shouldn't parse a point with invalid y coordinate %v", uncompressed)
+	}
+}
+
+func TestSchnorrPublicKey_SerializeFail(t *testing.T) {
+	pubkey := SchnorrPublicKey{}
+	_, err := pubkey.SerializeCompressed()
+	if err == nil {
+		t.Errorf("Zeroed public key isn't serializable")
+	}
+	_, err = pubkey.SerializeUncompressed()
+	if err == nil {
+		t.Errorf("Zeroed public key isn't serializable")
+	}
+}
+
+func TestBadPrivateKeyPublicKeyFail(t *testing.T) {
+	goodPrivateKey := fastGeneratePrivateKey(t)
+	goodPublicKey, err := goodPrivateKey.GenerateSchnorrPublicKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+	goodPublicKeyBackup := *goodPublicKey
+	goodPrivateKeyBackup := *goodPrivateKey
+	msg := fastGeneratePrivateKey(t).privkey
+	privkey := PrivateKey{}
+
+	_, err1 := privkey.GenerateSchnorrPublicKey()
+	_, err2 := privkey.SchnorrSign(msg)
+	_, err3 := ParsePrivateKey(privkey.privkey)
+	err4 := goodPrivateKey.Add(privkey.privkey)
+	err5 := goodPublicKey.Add(privkey.privkey)
+	if err1 == nil || err2 == nil || err3 == nil {
+		t.Errorf("A zeroed key is invalid")
+	}
+	if err4 != nil || err5 != nil {
+		t.Errorf("It should be possible to add zero to a key")
```

Comment:
Please check each error individually and add it to the error string here and everywhere. That way when the test fails we'll know exactly and immediately where it fails instead of having to connect a debugger.